### ISSUE TITLE
How safe is "safe hash"?

### ISSUE THREAD

**Issue by vt-alt** [cr-marcstevens/sha1collisiondetection#66]
> The library supports both an indicator flag that applications can check and act on, as well as a special safe-hash mode that returns the real SHA-1 hash when no collision was detected and a different safe hash when a collision was detected. Colliding files will have the same SHA-1 hash, but will have different unpredictable safe-hashes. This essentially enables protection of applications against SHA-1 collisions with no further changes in the application, e.g., digital signature forgeries based on SHA-1 collisions automatically become invalid.

Please clarify strength of the "safe hash"? Why it's not vulnerable to SHA-1 weaknesses?

**Comment by cr-marcstevens** [comment]
SHA-1 collision detection guaranteed detects collision attacks from 32 large families of collision attacks, the safe hash version processes any detected collision blocks twice.
Thus collision blocks are processed twice:
- the first time with the attack's intended chaining value
- the second time with the colliding chaining value.
The second time will break the collision, since the message blocks are different.
So SHA-1 colliding messages (constructed with any of the 32 attack families) don't collide anymore under SHA-1 safehash.

Basically to counter this, thus to construct a collision attack against safehash you need to:
- use a collision attack not of the 32 families detected
- OR modify the collision attack to cope with processing the collision block twice
- OR do a brute force collision attack
- OR be very lucky

**Comment by vt-alt** [comment]
Thanks for the answer. So (if I understood good) it as strong as it would be SHA-1 without known collision attacks

**Comment by cr-marcstevens** [comment]
No problem.

And this is almost as if it was fully secure.
The last three options above all have cost >= generic birthday attack cost.
Only the first option still could be faster, and likely is, than birthday.
However, as we picked the 32 best families, that is significantly more costly than the best attack.
I'm sure we conjectured the improved security level at the time somewhere, but I couldn't find it quickly atm.
